File: flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/template/JsInjectedGrandChild.java
/**
 * Copyright (C) 2024 Vaadin Ltd
 *
 * This program is available under Vaadin Commercial License and Service Terms.
 *
 * See {@literal <https://vaadin.com/commercial-license-and-service-terms>}  for the full
 * license.
 */
package com.vaadin.flow.uitest.ui.template;

import com.vaadin.flow.component.ClientCallable;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.HtmlImport;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
import com.vaadin.flow.templatemodel.TemplateModel;

@Tag("js-injected-grand-child")
@HtmlImport("frontend://com/vaadin/flow/uitest/ui/template/JsInjectedGrandChild.html")
@JsModule("JsInjectedGrandChild.js")
public class JsInjectedGrandChild extends PolymerTemplate<TemplateModel> {

    public JsInjectedGrandChild() {
        getElement().callJsFunction("greet");
        getElement().setProperty("bar", "foo");
    }

    @ClientCallable
    private void handleClientCall(String value) {
        getElement().setProperty("foo", value);
    }
}

File: flow-tests/test-root-context/src/main/frontend/JsInjectedGrandChild.js
import {PolymerElement} from '@polymer/polymer/polymer-element.js';
import {html} from '@polymer/polymer/lib/utils/html-tag.js';

class JsInjectedGrandChild extends PolymerElement {
  static get is() { return 'js-injected-grand-child' }

  greet() {
    this.$server.handleClientCall("bar");
  }

  static get template() {
    return html`
       <label id='foo-prop'>[[foo]]</label>
       <div id='prop'>[[bar]]</div>
    `;
  }
}

customElements.define(JsInjectedGrandChild.is, JsInjectedGrandChild);
